server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.coordination;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.druid.java.util.common.IAE;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This class keeps a bounded list of segment updates made on the server such as adding/dropping segments.
+ *
+ * Clients call addSegmentChangeRequest(DataSegmentChangeRequest) or addSegmentChangeRequests(List<DataSegmentChangeRequest>)
+ * to add segment updates.
+ *
+ * Clients call ListenableFuture<SegmentChangeRequestsSnapshot> getRequestsSince(final Counter counter) to get segment
+ * updates since given counter.
+ */
+public class SegmentChangeRequestHistory
+{
+  private static int MAX_SIZE = 1000;
+
+  private final int maxSize;
+
+  @GuardedBy("this")
+  private final List<Holder> changes;
+
+  @VisibleForTesting
+  @GuardedBy("waitingFutures")
+  final LinkedHashMap<CustomSettableFuture, Counter> waitingFutures;
+
+  private final ExecutorService singleThreadedExecutor;
+  private final Runnable resolveWaitingFuturesRunnable;
+
+  public SegmentChangeRequestHistory()
+  {
+    this(MAX_SIZE);
+  }
+
+  @VisibleForTesting
```

Comment:
I don't think this is only visible for testing.  Makes a ton of sense to me for this to be part of the public API.